diff -uNr linux-4.4.4/arch/ia64/include/asm/processor.h linux-4.4.4_changed/arch/ia64/include/asm/processor.h
--- linux-4.4.4/arch/ia64/include/asm/processor.h	2016-03-03 18:10:04.000000000 -0500
+++ linux-4.4.4_changed/arch/ia64/include/asm/processor.h	2016-03-27 20:04:29.203993919 -0400
@@ -202,6 +202,7 @@
 	unsigned int ptce_count[2];
 	unsigned int ptce_stride[2];
 	struct task_struct *ksoftirqd;	/* kernel softirq daemon for this CPU */
+	struct task_struct *my_ksoftirqd;	/* kernel softirq daemon for this CPU */
 
 #ifdef CONFIG_SMP
 	unsigned long loops_per_jiffy;
diff -uNr linux-4.4.4/include/linux/kthread.h linux-4.4.4_changed/include/linux/kthread.h
--- linux-4.4.4/include/linux/kthread.h	2016-03-03 18:10:04.000000000 -0500
+++ linux-4.4.4_changed/include/linux/kthread.h	2016-03-06 22:04:12.076339575 -0500
@@ -50,6 +50,8 @@
 void kthread_parkme(void);
 
 int kthreadd(void *unused);
+//added by Weiwei Jia for CS680 homework #6
+int __my_kernel_thread(void *data);
 extern struct task_struct *kthreadd_task;
 extern int tsk_fork_get_node(struct task_struct *tsk);
 
diff -uNr linux-4.4.4/init/main.c linux-4.4.4_changed/init/main.c
--- linux-4.4.4/init/main.c	2016-03-03 18:10:04.000000000 -0500
+++ linux-4.4.4_changed/init/main.c	2016-03-06 22:04:03.792133774 -0500
@@ -383,9 +383,74 @@
 
 static __initdata DECLARE_COMPLETION(kthreadd_done);
 
+//added by Weiwei Jia for CS680 homework #6
+#if 0
+static int my_kernel_kthread_create(void *data) {
+	printk("Weiwei Jia - my_kernel_kthread_create - %s\n", (char *) data);
+//	while(1) {
+//	}
+	return 0;
+}
+#endif
+
+//added by Weiwei Jia for CS680 homework #6
+//static void my_kernel_thread_create_1(void);
+//static void my_kernel_thread_create_2(void);
+
+#if 0
+//added by Weiwei Jia for CS680 homework #6
+static int m_k_t_do_something_2(void *data){
+	struct task_struct *curtask = current;
+	strcpy(curtask->comm, "Weiwei Jia: m_k_t_do_something_2");
+	set_task_state(curtask, TASK_RUNNING);
+	printk(KERN_NOTICE "Weiwei Jia: m_k_t_do_something_2 is about to be scheduled.\n");
+	local_irq_enable();
+	schedule();
+	local_irq_disable();
+	printk(KERN_NOTICE "Weiwei Jia: m_k_t_do_something_2 is now scheduled.\n");
+	return 0;
+}
+#endif
+
+//added by Weiwei Jia for CS680 homework #6
+static void my_kernel_thread_create_1(void){
+	int mypid;
+	printk(KERN_NOTICE "Weiwei Jia: Calling kernel_thread(wwj_kernel_thread_1)\n");
+	mypid = kernel_thread(&__my_kernel_thread, "wwj_ktd_1", CLONE_FS | CLONE_FILES | CLONE_SIGHAND);
+	printk(KERN_NOTICE "Weiwei Jia: wwj_kernel_thread_1 = %d\n", mypid);
+	return;
+}
+
+//added by Weiwei Jia for CS680 homework #6
+static void my_kernel_thread_create_2(void){
+	int mypid;
+	printk(KERN_NOTICE "Weiwei Jia: Calling kernel_thread(wwj_kernel_thread_2)\n");
+	mypid = kernel_thread(&__my_kernel_thread, "wwj_ktd_2", CLONE_FS | CLONE_FILES | CLONE_SIGHAND);
+	printk(KERN_NOTICE "Weiwei Jia: wwj_kernel_thread_2 = %d\n", mypid);
+	return;
+}
+
+//added by Weiwei Jia for CS680 homework #6
+static void __ps(void) {
+	struct task_struct *ps;
+	printk("Weiwei Jia - UID\tPID\tPPID\tCMD\n");
+	ps = current;
+	printk("Weiwei Jia -------------------------------------------------------------------\n");
+	read_lock(&tasklist_lock);
+	for_each_process(ps) {
+		printk("Weiwei Jia - %d\t%d\t%d\t%s\n", ps->loginuid.val, ps->pid, ps->parent->pid, ps->comm);
+	}
+	read_unlock(&tasklist_lock);
+	printk("Weiwei Jia -------------------------------------------------------------------\n");
+	return;
+}
+
 static noinline void __init_refok rest_init(void)
 {
 	int pid;
+	struct task_struct *my_thread1;
+	struct task_struct *my_thread2;
+	struct task_struct *my_thread3;
 
 	rcu_scheduler_starting();
 	smpboot_thread_init();
@@ -402,6 +467,29 @@
 	rcu_read_unlock();
 	complete(&kthreadd_done);
 
+	//added by Weiwei Jia
+	printk(KERN_NOTICE "Weiwei Jia - CS680 homework #6, start ...\n");
+	printk(KERN_NOTICE "Weiwei Jia: two threads are about to be created by kernel_thread func.\n");
+#if 1
+	my_kernel_thread_create_1();
+	my_kernel_thread_create_2();
+	printk(KERN_NOTICE "Weiwei Jia: two threads are created by kernel_thread func.\n");
+
+	printk(KERN_NOTICE "Weiwei Jia: three threads are about to be created by kthread_create func.\n");
+	my_thread1 = kthread_create(&__my_kernel_thread, "K_C_1", "WeiweiJ-KC1");
+	wake_up_process(my_thread1);
+	my_thread2 = kthread_create(&__my_kernel_thread, "K_C_2", "WeiweiJ-KC2");
+	wake_up_process(my_thread2);
+	my_thread3 = kthread_create(&__my_kernel_thread, "K_C_3", "WeiweiJ-KC3");
+	wake_up_process(my_thread3);
+	printk(KERN_INFO "Weiwei Jia - This is %s\n", my_thread1->comm);
+	printk(KERN_INFO "Weiwei Jia - This is %s\n", my_thread2->comm);
+	printk(KERN_INFO "Weiwei Jia - This is %s\n", my_thread3->comm);
+	printk(KERN_NOTICE "Weiwei Jia: three threads are created by kthread_create func.\n");
+#endif
+	__ps();
+	printk(KERN_NOTICE "Weiwei Jia - CS680 homework #6, end ...\n");
+
 	/*
 	 * The boot idle thread must execute schedule()
 	 * at least once to get things moving:
@@ -929,6 +1017,8 @@
 
 static noinline void __init kernel_init_freeable(void);
 
+
+
 static int __ref kernel_init(void *unused)
 {
 	int ret;
diff -uNr linux-4.4.4/kernel/kthread.c linux-4.4.4_changed/kernel/kthread.c
--- linux-4.4.4/kernel/kthread.c	2016-03-03 18:10:04.000000000 -0500
+++ linux-4.4.4_changed/kernel/kthread.c	2016-03-06 22:04:23.500623383 -0500
@@ -499,6 +499,40 @@
 }
 EXPORT_SYMBOL(kthread_stop);
 
+//added by Weiwei Jia for CS680 homework #6
+int __my_kernel_thread(void *data){
+	struct task_struct *curtask = current;
+	set_task_comm(curtask, (char *) data);
+	ignore_signals(curtask);
+	set_cpus_allowed_ptr(curtask, cpu_all_mask);
+	set_mems_allowed(node_states[N_MEMORY]);
+	//__set_task_state(curtask, TASK_RUNNING);
+	printk(KERN_NOTICE "Weiwei Jia: %s is being scheduled.\n", (char *) data);
+	for (;;) {
+		set_current_state(TASK_INTERRUPTIBLE);
+		if (list_empty(&kthread_create_list))
+			schedule();
+		__set_current_state(TASK_RUNNING);
+
+		spin_lock(&kthread_create_lock);
+		while (!list_empty(&kthread_create_list)) {
+			struct kthread_create_info *create;
+
+			create = list_entry(kthread_create_list.next,
+					    struct kthread_create_info, list);
+			list_del_init(&create->list);
+			spin_unlock(&kthread_create_lock);
+
+			create_kthread(create);
+
+			spin_lock(&kthread_create_lock);
+		}
+		spin_unlock(&kthread_create_lock);
+	}
+
+	return 0;
+}
+
 int kthreadd(void *unused)
 {
 	struct task_struct *tsk = current;
diff -uNr linux-4.4.4/kernel/softirq.c linux-4.4.4_changed/kernel/softirq.c
--- linux-4.4.4/kernel/softirq.c	2016-03-03 18:10:04.000000000 -0500
+++ linux-4.4.4_changed/kernel/softirq.c	2016-03-27 20:03:47.402879210 -0400
@@ -56,6 +56,8 @@
 static struct softirq_action softirq_vec[NR_SOFTIRQS] __cacheline_aligned_in_smp;
 
 DEFINE_PER_CPU(struct task_struct *, ksoftirqd);
+//added by Weiwei Jia
+DEFINE_PER_CPU(struct task_struct *, my_ksoftirqd);
 
 const char * const softirq_to_name[NR_SOFTIRQS] = {
 	"HI", "TIMER", "NET_TX", "NET_RX", "BLOCK", "BLOCK_IOPOLL",
@@ -72,9 +74,14 @@
 {
 	/* Interrupts are disabled: no need to stop preemption */
 	struct task_struct *tsk = __this_cpu_read(ksoftirqd);
+	//added by Weiwei Jia
+	struct task_struct *my_tsk = __this_cpu_read(my_ksoftirqd);
 
 	if (tsk && tsk->state != TASK_RUNNING)
 		wake_up_process(tsk);
+	
+	if (my_tsk && my_tsk->state != TASK_RUNNING)
+		wake_up_process(my_tsk);
 }
 
 /*
@@ -651,6 +658,13 @@
 	return local_softirq_pending();
 }
 
+//added by Weiwei Jia
+static int my_ksoftirqd_should_run(unsigned int cpu)
+{
+	printk("999dbg, Enter my ksoftirqd should run!\n");
+	return local_softirq_pending();
+}
+
 static void run_ksoftirqd(unsigned int cpu)
 {
 	local_irq_disable();
@@ -667,6 +681,26 @@
 	local_irq_enable();
 }
 
+
+//added by Weiwei Jia
+static void my_run_ksoftirqd(unsigned int cpu)
+{
+	printk("999dbg, Enter my run ksoftirqd\n");
+	local_irq_disable();
+	if (local_softirq_pending()) {
+		/*
+		 * We can safely run softirq on inline stack, as we are not deep
+		 * in the task stack here.
+		 */
+		__do_softirq();
+		local_irq_enable();
+		cond_resched_rcu_qs();
+		return;
+	}
+	local_irq_enable();
+}
+
+
 #ifdef CONFIG_HOTPLUG_CPU
 /*
  * tasklet_kill_immediate is called to remove a tasklet which can already be
@@ -740,6 +774,22 @@
 	return NOTIFY_OK;
 }
 
+//added by Weiwei Jia
+static int my_cpu_callback(struct notifier_block *nfb, unsigned long action,
+			void *hcpu)
+{
+	printk("999dbg, my cpu callback!\n");
+	switch (action) {
+#ifdef CONFIG_HOTPLUG_CPU
+	case CPU_DEAD:
+	case CPU_DEAD_FROZEN:
+		takeover_tasklets((unsigned long)hcpu);
+		break;
+#endif /* CONFIG_HOTPLUG_CPU */
+	}
+	return NOTIFY_OK;
+}
+
 static struct notifier_block cpu_nfb = {
 	.notifier_call = cpu_callback
 };
@@ -751,6 +801,30 @@
 	.thread_comm		= "ksoftirqd/%u",
 };
 
+//added by Weiwei Jia
+static struct notifier_block my_cpu_nfb = {
+	.notifier_call = my_cpu_callback
+};
+
+static struct smp_hotplug_thread my_softirq_threads = {
+	.store			= &my_ksoftirqd,
+	.thread_should_run	= my_ksoftirqd_should_run,
+	.thread_fn		= my_run_ksoftirqd,
+	.thread_comm		= "my_ksoftirqd/%u",
+};
+
+static __init int my_spawn_ksoftirqd(void)
+{
+	printk("999dbg, Enter my spawn ksoftirqd\n");
+	register_cpu_notifier(&my_cpu_nfb);
+
+	BUG_ON(smpboot_register_percpu_thread(&my_softirq_threads));
+
+	return 0;
+}
+early_initcall(my_spawn_ksoftirqd);
+//ended by Weiwei Jia
+
 static __init int spawn_ksoftirqd(void)
 {
 	register_cpu_notifier(&cpu_nfb);
